### PR TITLE
[cli] Log delta in `vc env pull`

### PR DESCRIPTION
**Pull request by MatthewStanciu**
This PR adds a new message to the end of `vc env pull` which shows delta of what was added, changed, and removed.

Example: you have a local `.env` file in a directory linked to a Vercel project. You change `TEST_1`, remove `TEST_2` via `vc env rm TEST_2`, and add `TEST_3` via `vc env add`. Then you run `vc env pull`. Here's what you will see:

```
> Updating existing .env file
> Downloading `development` Environment Variables for Project name
✅  Updated .env file

+ TEST_3
~ TEST_1
- TEST_2
```

Screen Shot 2022-07-20 at 11 09 56 AM

#### Tests

- [x] The code changed/added as part of this PR has been covered with tests
- [x] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/cli/src/commands/env/pull.ts
Comment (EndangeredMassa) [inline packages/cli/src/commands/env/pull.ts:145]: If `createEnvObject` throws an error, it will end up returning `undefined`. When it gets down here, there could have been actual changes, but we'll log here that there were none. In this case, I think it's better to either say nothing or be clear that we couldn't detect the changes, but the latest values were updated.

File: packages/cli/src/commands/env/pull.ts
Comment (EndangeredMassa) [inline packages/cli/src/commands/env/pull.ts]: Should this check that there are more items as well? If there are exactly 3 items, will this render all 3 items plus an ellipses?

Can we add a test for boundary cases like this?

File: packages/cli/src/commands/env/pull.ts
Comment (EndangeredMassa) [inline packages/cli/src/commands/env/pull.ts]: These functions related to env diffing could be better organized into a separate file, in my opinion.

File: packages/cli/src/commands/env/pull.ts
Comment (EndangeredMassa) [inline packages/cli/src/commands/env/pull.ts]: I think it makes sense that `~` would mean "changed", but not immediately. I think a user may not know what this means, especially if their first interactions with this output only shows one variable changed.

Should we log a `(changed)` suffix (maybe dimmed) at the end of these lines? Or can we show a brief legend without taking up too much space?

Comment (MatthewStanciu) [reply packages/cli/src/commands/env/pull.ts]: I'll explore a legend, though not quite sure yet how to make it look good & not feel awkward. Big fan of the `(changed)` suffix.

File: packages/cli/src/commands/env/pull.ts
Comment (MatthewStanciu) [reply packages/cli/src/commands/env/pull.ts]: Will add a test for the ellipsis case. 👍 

The intention is to render 3 items plus an ellipsis. The thinking here is that if someone made some major changes to their env variables, I wouldn't want the diff to take up too much of their screen.

File: packages/cli/src/commands/env/pull.ts
Comment (MatthewStanciu) [reply packages/cli/src/commands/env/pull.ts]: Added a `Changed:` prefix, but not super confident in this

Comment (EndangeredMassa) [reply packages/cli/src/commands/env/pull.ts]: I think the `Changed:` prefix is fine, but not necessary. I think the issue with `~` still stands, though.

File: packages/cli/src/commands/env/pull.ts
Comment (EndangeredMassa) [reply packages/cli/src/commands/env/pull.ts]: Did you intentionally remove the ellipses code? I think both approaches (with or without truncation) is fine, but wanted to make sure this was the intent. If so, the fixture `env-pull-delta-ellipsis` is still in the PR.

If not, the test you added was removed. You'll want to restore that. I also suggest adding a test for exactly 3 items not showing an ellipses.

Comment (MatthewStanciu) [reply packages/cli/src/commands/env/pull.ts]: Ah, thanks for catching. I intentionally removed the ellipsis in favor of displaying changed items in a list. I think it didn't really make sense to show each variable on its own line, and I think showing them in a list makes more sense.

Will remove the fixture.

Comment (EndangeredMassa) [reply packages/cli/src/commands/env/pull.ts]: Oh, yeah, I like listing them a lot

File: packages/cli/src/commands/env/pull.ts
Comment (MatthewStanciu) [reply packages/cli/src/commands/env/pull.ts]: Thinking about it more, I'm not convinced this is an issue. I think there is enough context as to what the tilde means: there is now a "changed" prefix which gives a clue, especially if there's only one edited variable, and the yellow color indicates (at least to me) that the variable wasn't added or removed, but somewhere in between—AKA edited.

I think a legend, or more explicit context, will take up a lot of space, and is the sort of thing that may be useful to see once but will be annoying to see every time after that.

I would be happy to ship this as is. Though I think it's worth bringing this to the rest of the team or to `#design` / `#engineering` for additional feedback.

File: packages/cli/src/commands/env/pull.ts
Comment (Ethan-Arrowood) [inline packages/cli/src/commands/env/pull.ts]: can you add a test for this part

File: packages/cli/src/util/env/diff-env-files.ts
Comment (Ethan-Arrowood) [inline packages/cli/src/util/env/diff-env-files.ts]: never re-assigned so these can be `const`
```suggestion
  const added = [];
  const changed = [];
```

File: packages/cli/src/util/env/diff-env-files.ts
Comment (Ethan-Arrowood) [inline packages/cli/src/util/env/diff-env-files.ts]:
```suggestion
```
remove this and just declare as constant on line when its assigned

File: packages/cli/src/util/env/diff-env-files.ts
Comment (Ethan-Arrowood) [inline packages/cli/src/util/env/diff-env-files.ts]:
```suggestion
  const removed = Object.keys(oldEnv);
```

File: packages/cli/src/util/env/diff-env-files.ts
Comment (Ethan-Arrowood) [inline packages/cli/src/util/env/diff-env-files.ts]:
```suggestion
  const { added, changed, removed } = findChanges(oldEnv, newEnv);
```
this should be caught by a lint rule but preference for using `const` when something is not re-assigned

File: packages/cli/src/util/parse-env.ts
Comment (Ethan-Arrowood) [inline packages/cli/src/util/parse-env.ts:12]: surprised typescript isn't catching this... make sure to update the `env` function parameter type to include `string`

File: packages/cli/src/util/parse-env.ts
Comment (Ethan-Arrowood) [inline packages/cli/src/util/parse-env.ts]:
```suggestion
export const parseEnv = (env?: string | string[] | Dictionary<string>) => {
```

File: packages/cli/src/commands/env/pull.ts
Comment (EndangeredMassa) [inline packages/cli/src/commands/env/pull.ts]: Should we say anything about failing in this case? I think we should only say `the .env file was updated`.

**Review by EndangeredMassa — APPROVED** [review]
I'm good with this version, but you may want to wait for confirmation from others who had suggestions on this before merging.

File: packages/cli/src/commands/env/pull.ts
Comment (MatthewStanciu) [reply packages/cli/src/commands/env/pull.ts]: I think you're right. The goal is to make the user aware that something went wrong if they were expecting a diff, but this is an awkward interaction. Will remove 👍

File: packages/cli/src/util/parse-env.ts
Comment (styfle) [inline packages/cli/src/util/parse-env.ts]: This comment should be on the function below, not the import

File: packages/cli/src/util/parse-env.ts
Comment (styfle) [inline packages/cli/src/util/parse-env.ts]:
```suggestion
import type { Dictionary } from '@vercel/client';
```

File: packages/cli/test/unit/commands/env.test.ts
Comment (EndangeredMassa) [inline packages/cli/test/unit/commands/env.test.ts]: I think we should still have a test for this case. The output should say that the `.env` file was updated, but it should not show a diff.

**Review by EndangeredMassa — APPROVED** [review]

File: packages/cli/test/unit/commands/env.test.ts
Comment (MatthewStanciu) [reply packages/cli/test/unit/commands/env.test.ts]: I re-added the test, but I think it's hard to test for this. In every case where the env file was updated, it outputs "Updated <env name> file". In most cases, it then also outputs "No changes found", but in the case where it can't parse the file, it doesn't output that. So we want to test that something _doesn't_ output, which I'm not sure is possible right now.

Comment (EndangeredMassa) [reply packages/cli/test/unit/commands/env.test.ts]: Ah, right, because our output mocking changed. This will at least make sure there's no error in this case.

**Review by EndangeredMassa — APPROVED** [review]

**Review by cb1kenobi — APPROVED** [review]

**Review by TooTallNate — APPROVED** [review]

File: packages/cli/src/util/env/diff-env-files.ts
Comment (styfle) [inline packages/cli/src/util/env/diff-env-files.ts]: What is the value its checking for here? Is this basically the same

```suggestion
    if (oldEnv[key] === undefined) {
```

Or are you expecting null value in addition to undefined?

I'm also curious how to tell the difference between undefined and empty string. I don't think we allow an undefined value today.

Comment (MatthewStanciu) [reply packages/cli/src/util/env/diff-env-files.ts]: I'm just checking for undefined. This is better, thank you

File: packages/cli/src/util/env/diff-env-files.ts
Comment (styfle) [inline packages/cli/src/util/env/diff-env-files.ts]:
```suggestion
  return arr.sort().map(item =>`${prefix} ${item}`).join('\n') + '\n'
```

Nit pick: this can be simplified

Comment (MatthewStanciu) [reply packages/cli/src/util/env/diff-env-files.ts]: Thank you 🙏

**Review by styfle — APPROVED** [review]
Looks good, thanks!

File: packages/cli/src/util/env/diff-env-files.ts
Comment (styfle) [inline packages/cli/src/util/env/diff-env-files.ts]:
```suggestion
  return arr.sort().map(item =>`${prefix} ${item}`).join('\n') + '\n';
```

File: packages/cli/src/util/env/diff-env-files.ts
Comment (styfle) [inline packages/cli/src/util/env/diff-env-files.ts]:
```suggestion
  return arr.sort().map(item =>`${prefix} ${item}`).join('\n') + '\n';
```

File: packages/cli/src/util/env/diff-env-files.ts
Comment (styfle) [inline packages/cli/src/util/env/diff-env-files.ts]:
```suggestion
  // Originally authored by Tyler Waters under MIT License: https://github.com/tswaters/env-file-parser/blob/f17c009b39da599380e069ee72728d1cafdb56b8/lib/parse.js
  // https://github.com/tswaters/env-file-parser/blob/f17c009b39da599380e069ee72728d1cafdb56b8/LICENSE
```

**Pull request merged**
